pkg/graphviz/traceflow.go

```diff
@@ -204,6 +169,82 @@ func genOutput(g *graphviz.Graphviz, graph *cgraph.Graph, isSingleCluster bool)
 	return str
 }
 
+func genSubGraph(graph *cgraph.Graph, result *opsv1alpha1.NodeResult, firstNodeName string, dir cgraph.DirType, addNodeNum int) []*cgraph.Node {
+	var nodes []*cgraph.Node
+	obs := result.Observations
+
+	// Show the name of cluster.
+	if len(result.Node) > 0 {
+		graph.SetLabel(result.Node)
+		if dir == cgraph.ForwardDir {
+			graph.SetLabelJust(cgraph.LeftJust)
+		} else {
+			graph.SetLabelJust(cgraph.RightJust)
+		}
+	}
+
+	// Construct the first node. Show it only if we know the name of it.
+	node := createEndpointNodeWithDefaultStyle(graph, firstNodeName)
+	nodes = append(nodes, node)
+	if len(firstNodeName) > 0 {
+		node.SetColor(grey)
+		node.SetFillColor(lightGrey)
+		node.SetStyle(cgraph.BoldNodeStyle + "," + cgraph.FilledNodeStyle)
+	} else {
+		node.SetStyle("invis")
+	}
+
+	// Reorder the observations according to the direction of edges.
+	if dir == cgraph.BackDir {
+		for i := len(obs)/2 - 1; i >= 0; i-- {
+			opp := len(obs) - 1 - i
+			obs[i], obs[opp] = obs[opp], obs[i]
+		}
+	}
+
+	// Draw the actual observations of traceflow.
+	for _, o := range obs {
+		// Construct node and edge.
+		nodeName := fmt.Sprintf("%s_%d", graph.Name(), len(nodes))
+		node := createNodeWithDefaultStyle(graph, nodeName)
+		node.SetLabel(string(o.Component))
+		nodes = append(nodes, node)
+		if len(nodes) > 1 {
+			edgeName := fmt.Sprintf("%s_%d", graph.Name(), len(nodes))
+			edge := createEdgeWithDefaultStyle(graph, edgeName, nodes[len(nodes)-2], nodes[len(nodes)-1])
+			edge.SetDir(dir)
+			if len(nodes) == 2 {
```

Comment:
So we will always adjust the edge between the first two nodes to make sure the two graphs have the same length?
If so, please add some comments here.

Reply:
Comment added.